### PR TITLE
[scudo] Add hooks to mark the range of realloc

### PR DESCRIPTION
**Pull request by ChiaHungDuan**
`realloc` may involve both allocation and deallocation. Given that the reporting the events is not atomic and which may lead the hook user to a false case that the double-use pattern happens. In general, this can be resolved on the hook side. To alleviate the task of handling it, we add two new hooks to mark the range so that the hook user can combine those calls together.

### CODE REVIEW

**Comment by cferris1000** [comment]
Fabio, since you are the person who is most interested in this change, can you indicate how this solves your problem? For example, it's not clear how a snapshot from a profiler potentially being slightly wrong if it gets caught between the deallocate hook and the allocate hook in the realloc code is a problem. I'm not familiar with the profilers and how they are used by your team, so this could be something that occurs frequently. If you are using a sampler profiler, there is always the possibility of a snapshot being incorrect since you could get caught before the hook is called but after the operation has completed depending on how the profiler is coded.

I'm not opposed to this change, I just want to make sure we are adding code that solves a real problem, and not adding code for a corner case that is not an issue for any current profilers.

**Comment by cferris1000** [comment]
To add concrete numbers, I looked through the memory traces we have taken on Android devices, and reallocs tend to be about 1%-2% or 3%-4% of the operations. The highest I saw was 6%, which seems to be a particular app. So the chance of hitting this issue is already relatively low, but it's possible that you are seeing a different rate that is higher than on Android.

**Review by fabio-d — APPROVED** [review]
Thanks @ChiaHungDuan and @cferris1000. This new revision solves my problem of tracking memory precisely even while it's being reallocated.

File: compiler-rt/lib/scudo/standalone/wrappers_c.inc [hunk 1]
```diff
@@ -27,15 +27,21 @@ static void reportDeallocation(void *ptr) {
...
 }
-static void reportReallocBegin(void *old_ptr) {
-  if (SCUDO_ENABLE_HOOKS)
-    if (__scudo_realloc_begin_hook)
-      __scudo_realloc_begin_hook(old_ptr);
+static void reportReallocAllocation(void *old_ptr, void *new_ptr, size_t size) {
+  if (SCUDO_ENABLE_HOOKS) {
+    if (__scudo_realloc_allocate_hook)
+      __scudo_realloc_allocate_hook(old_ptr, new_ptr, size);
+    else
```
[the remark this answers is not in the document]
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/wrappers_c.inc:34]: if `new_ptr` == nullptr, it'll be handled differently. See the caller for the detail.

The `else if (__scudo_allocate_hook)` has fixed.

File: compiler-rt/lib/scudo/standalone/wrappers_c.inc [hunk 2]
```diff
@@ -27,15 +27,21 @@ static void reportDeallocation(void *ptr) {
...
 }
-static void reportReallocEnd(void *old_ptr) {
-  if (SCUDO_ENABLE_HOOKS)
-    if (__scudo_realloc_end_hook)
-      __scudo_realloc_end_hook(old_ptr);
+static void reportReallocDeallocation(void *old_ptr) {
+  if (SCUDO_ENABLE_HOOKS) {
+    if (__scudo_realloc_deallocate_hook)
+      __scudo_realloc_deallocate_hook(old_ptr);
+    else
```
[the remark this answers is not in the document]
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/wrappers_c.inc:42]: Done

File: compiler-rt/lib/scudo/standalone/wrappers_c.inc (hunk 1 above)
[the remark this answers is not in the document]
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/wrappers_c.inc:34]: BTW, added a DCHECK for the new_ptr != nullptr

File: compiler-rt/lib/scudo/standalone/include/scudo/interface.h [hunk 3]
```diff
@@ -20,6 +20,20 @@ __attribute__((weak)) const char *__scudo_default_options(void);
 __attribute__((weak)) void __scudo_allocate_hook(void *ptr, size_t size);
 __attribute__((weak)) void __scudo_deallocate_hook(void *ptr);
 
+// `realloc` involves both deallocation and allocation but they are not reported
+// atomically. In one specific case which may keep taking a snapshot right in
+// the middle of `realloc` reporting the deallocation and allocation, it may
+// confuse the user by the missing memory from `realloc`. To alleviate that
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/include/scudo/interface.h:26]: by the missing -> by missing

File: compiler-rt/lib/scudo/standalone/include/scudo/interface.h [hunk 4]
```diff
@@ -20,6 +20,20 @@ __attribute__((weak)) const char *__scudo_default_options(void);
...
+// case, define the two `realloc` hooks to get the knowledge of the bundled hook
+// calls. This hooks are optional and only used when you are pretty likely to
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/include/scudo/interface.h:28]: This -> These

and only used when you are pretty likely to -> and should only be used when a hooks user wants to track reallocs more closely.

or something similar.

File: compiler-rt/lib/scudo/standalone/include/scudo/interface.h (hunk 3 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/include/scudo/interface.h:26]: Done

File: compiler-rt/lib/scudo/standalone/tests/wrappers_c_test.cpp [hunk 5]
```diff
@@ -73,6 +79,28 @@ __attribute__((visibility("default"))) void __scudo_allocate_hook(void *Ptr,
...
+  // Verify that __scudo_realloc_deallocate_hook is called first and set the
+  // right pointer.
+  EXPECT_EQ(OldPtr, RC.DeallocPtr);
+  RC.AllocPtr = NewPtr;
+  RC.Size = Size;
+
+  // Note that this is only used for testing. In general, only one pair of hooks
+  // will be invoked in `realloc`. if __scudo_realloc_*_hook are not defined,
+  // it'll call the general hooks only. To make the test easier, we call the
+  // general one here so that either case (wether __scudo_realloc_*_hook are
```
Comment (cferris1000) [inline compiler-rt/lib/scudo/standalone/tests/wrappers_c_test.cpp:93]: wether -> whether

**Review by cferris1000 — COMMENTED** [review]
One small typo.

File: compiler-rt/lib/scudo/standalone/tests/wrappers_c_test.cpp (hunk 5 above)
Comment (ChiaHungDuan) [reply compiler-rt/lib/scudo/standalone/tests/wrappers_c_test.cpp]: Done

**Review by cferris1000 — APPROVED** [review]
LGTM.

**Pull request merged**